### PR TITLE
Fix for PR #1149

### PR DESCRIPTION
**Pull request by erezsh**

### CODE REVIEW

File: lark/tree_templates.py [hunk 1]
```diff
@@ -49,15 +53,21 @@ def _get_tree(self, template: TreeOrCode) -> Tree[str]:
...
         return Template(template, conf=self)
 
-    def _match_tree_template(self, template: TreeOrCode, tree: TreeOrCode) -> Optional[Dict[str, TreeOrCode]]:
+    def _match_tree_template(self, template: TreeOrCode, tree: Branch) -> Optional[MatchResult]:
+        """Returns dict of {var: match} if found a match, else None
+        """
         template_var = self.test_var(template)
         if template_var:
+            if not isinstance(tree, Tree):
+                raise TypeError(f"Template variables can only match Tree instances. Not {tree}")
```
Comment (MegaIng) [inline lark/tree_templates.py:70]: This should be either `type(tree)` or `{tree!r}`

File: lark/tree_templates.py [hunk 2]
```diff
@@ -9,14 +9,18 @@
...
 
 class TemplateConf:
     """Template Configuration
 
     Allows customization for different uses of Template
+
+    parse() must return a Tree instance.
     """
 
     def __init__(self, parse=None):
```
Comment (MegaIng) [inline lark/tree_templates.py:26]: This could be annotated with `Callable[[str], Tree]`

**Review by MegaIng — CHANGES_REQUESTED** [review]

File: lark/tree_templates.py (hunk 2 above)
Comment (erezsh) [reply lark/tree_templates.py:26]: I agree, but there's a lot of "could be" to add there. We can do that in a different PR.

File: lark/tree_templates.py (hunk 1 above)
Comment (erezsh) [reply lark/tree_templates.py]: Fixed.

**Pull request merged**